### DISCUSSION TITLE
State is getting confused for multiple instances of the same component type

### DISCUSSION THREAD

**Discussion by EverettMcKay** [MithrilJS/mithril.js#2995] [Q&A]
Suppose I have several different components with this basic structure:
```
<Container>
	<Body />
	<Feedback />
</Container>
```
where Feedback is displayed over the entire Container when the user is done interacting with the Body. (By using position: absolute and z-index:2)
In almost all cases, this works as expected. For example, the general case might be this:
```
	<ComponentType1 />
	<ComponentType2 />
	<ComponentType3 />
```
This works. However, this doesn’t:
```
	<ComponentType1-1 />
	<ComponentType1-2 />
```
In other words, if I have two instances of the same component type, things break down. What happens? Feedback for ComponentType1-2 is displayed on top of ComponentType1-1.
Code-wise, Feedback is set as the results of a timer. Something like this:
```
    state.feedbackTimer = setTimeout(() => {
      state.feedbackShown = true;
      m.redraw();
    }, 2000);
```
But somehow, the state getting changed is for the wrong component.

Is there any reason why multiple instances of a Mithril component would get their states confused like this? And if so, is there a way to prevent this confusion?

**Comment by dead-claudia** [comment]
What's your component code look like? This so far only makes sense if your view state is managed outside of your component (a massive anti-pattern, by the way).

**Reply by EverettMcKay** [reply]
I'm using this shape:

```
export const Component = (): IComponent => ({
  tag: "component",
  attrs: () => {},
  state: () => {},
  oncreate: () => { },
  oninit: () => { },
  view: () => ( <div>Hello</div>),
});
```

**Reply by dead-claudia** [reply]
@EverettMcKay [Are you using that as `m(Component())`?](https://mithril.js.org/components.html#define-components-statically,-call-them-dynamically)

**Comment by JAForbes** [comment]
Is `state` a global or module scoped variable?  Do you mean `vnode.state` ?

**Comment by EverettMcKay** [comment]
Here is the gist of what I am doing (skipping the styling):
```
interface IAttrs {
  name: string;
}
interface IFeedbackAttrs {
  show: boolean;
}
interface IState {
  feedbackTimer: number;
  feedbackShown: boolean;
}

const Feedback = (): IComponent => ({
  tag: "feedback",
  attrs: () => {},
  state: () => {},
  view: (v: Vnode<IFeedbackAttrs, any>) => (v.attrs.show ? <div>Thank you!</div> : null),
});

const Component = (): IComponent => ({
  tag: "component",
  oninit: (v: Vnode<IAttrs, IState>) => {
    v.state.feedbackTimer = 0;
    v.state.feedbackShown = false;
  },
  attrs: () => {},
  state: () => {},
  view: (v: Vnode<IAttrs, IState>) => (
    <div>
      <button
        onclick={(ev: Event) => {
          ev.preventDefault();
          v.state.feedbackTimer = setTimeout(() => {
            v.state.feedbackShown = true;
            m.redraw();
          }, 2000);
        }}>{`Hello, ${v.attrs.name} `}</button>
      <Feedback show={v.state.feedbackShown} />
    </div>
  ),
});

const HomePage = (): IComponent => ({
  tag: "home",
  attrs: () => {},
  state: () => {},
  view: () => (
    <>
      <Component name="Bob" />
      <Component name="Alice" />
    </>
  ),
});

m.route(document.body, "/home", {
  "/home": HomePage,
});
```

The problem is that clicking on the button in Alice displays the thank you in Bob.

**Reply by dead-claudia** [reply]
[That doesn't replicate for me.](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4A9ACoABAAF+ADylZZCgGIAnDAHMpAcmS6pEsQB00ZoXGJTVMGABMARhmoBrKQF4pACgCUngD4fYDMpKWJtRCkTcDsnF1cYilCpDGJidTgov0CpYHYUqzSYbP8PIPzktDCANwgYAHdsmrKg7xr8NIy4fDhCWgapAH4pAB57CBqAgBVCDDR3AE9aAFcAQlGxCampKLQVqChfKtZfAG4zC3orKQBhWiwAB3oRTx9W4JSIrSiYuieXgwkil6BA0BBiM0PiFqmEpB0iox8GA4s43NMIDh1G8AAwXWG1XoRJEohxo1wAZX6DWqXjAGCgcBg+LCBVhXUypVylUKxJK725bNq9SaPhauW8KTC40mASlcNGjhW6Xo8rhUno1CgEDcHmA3hg4vKeTV6qkhvwj3UhpEABEYPSDsQ-CyzeqEXzkaiEhisW8mcRfTBVs6csaYW63R7il6yQkqQNaeF1Ctmaa3Vh8Nb7JoGi703D2FIAEw4svnAusVgBYAAAwAEjBDrQKFIACTADocnqYHCsKS11ibJUqtBygkK2xxtxSPoDPXRknetwJmn9sTj9WbbabqTHMynfFXNA3esPGAABW0MDeYbaEfCkWiIH6OGB7PSnIF4aFs75XJ-KphUaAC2nlUZdy3e4ATQV5exgDwYgAIVoRwYikDdTVGaDnlghgpHgxCQAAQW1agYHQzCJ02Xd9zQQ9LjQTN1BDGBvHsWhqBWHAGHwRxaHsRZWxiMRXwoygTVhESxJiKIzxwK8tBgE5zkoGgHkeaAYHUPBnEcJs1KZWBqGICBrjwYsADZEGLABONgOBAeC8Hwag4AEdThGYHg2AAXSobUFgQFBOBwPAsAhQh1GgNSVnUcgeBIYhHiyMQxBWNBHlcLRXIeMQIuIKLoBkYt8FKgBGHF8si6KoHwCK0Hwfg1OIRZHm4EA4GoaLHlEVhfNYIA) Is there something else missing?

(All I did was drop the type annotations in that and plumb it through Babel instead of TypeScript, by the way.)

**Reply by EverettMcKay** [reply]
My apologies... My goal with the code is to summarize my approach, hoping there was an obvious, known problem regarding component state.

I will work on a code sample that reproduces the problem and repost. (I'm not sure what is missing.)

**Reply by dead-claudia** [reply]
@EverettMcKay My usual process is just start with whatever the buggy code is, and then just (without committing anything) remove chunks of code and check if it continues to reproduce. It's a bit slower, but it gets me to a minimal repro 100% of the time. (I've even narrowed down [a few](https://github.com/microsoft/TypeScript/issues/50952) [very subtle](https://github.com/microsoft/TypeScript/issues/20558) [TS bugs](https://github.com/microsoft/TypeScript/issues/18299) this way.)

**Comment by JAForbes** [comment]
I made the smallest change to your example just to get it to work in a code playground.  And I can't reproduce the bug: 

[link](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4hcYgAIAYjBgATAEYZqAa0kBeSQAoAlJoB8O4AB00kycQwBzRJJPg5SlaocUzFjMWIAnOHb1DSWB2D0kJLxgA-Q0jEPdzSQA3CBgAdwCkmKNtJPwvXzh8OEJaNMkAfkksbQByWopagBVCDDR1AE9aAFcAQlr9OzRuqChdBNZdAG4zM3EpAGFaLAAHehFNHWzjMKtbexpltbQRNzD6CDQIYkzt00SLPIjGfDAnZTUmiBwfTYAGGYPZLFKwvN4KD6qADKpTS5i0YAwUDgMEBFlCiQKfmiQXiYWeUS2uIxj1SGR0WSCYQsNXq1Iskgo1Tqim63notXpDMZwUk9GoUAgagCMEpsWCXO5klF+BWPlFIgAIjBESNiHo0VLuU9QTBXu8XF8fpsUcQjTAeurAuL7lqtTrIvqIS4YWV4ZYfN1UZKtVh8PL5D4MGkNT6GexJAAmP4x6Zh1isMNM2oACRgo1okgawKxRUwOEl+klTJqsmdaiZwHCsLsDrBBrUrrhkkm9N0ZkmgLm9AkkhTyxgAAUbDBNtacrbLDY7A5SgXKGFczibSTwrrl3FVyl0hudPTaZygRYS9olqt1gxK5J84TagAhWiKWototHnk1M-HERXm92WoAQUFagYGfSZ7CBdtEnGDs4zQMw-R8S0YG0eRaGobocAYfBFFoeQOiZBwAHo5xgNwJUSIiSIcOx+xwYdrBgCZYMoQ5VmgGAfCQYZRioFFYGoYgIB7PA-kQP4AFoABYADYxLYDgQBvPB8GoOABEOYRmB4NgAF0qEFdoEBQTgcDwLBrkIHxoBY7ofHIHgSGIFZ-EIwjujQFZVGsFTlkI8ziEs6AAAFI3wUKAEY-j8iyrKgfBzLQfB+BY4gOhWbgQDgagrJWURWB01ggA)

But also there's a few things in your example that seem odd.

1. `state: () => {},` and ` attrs: () => {},` what is the intention for those function definitions?  You don't need that.
2. What is the intention behind manually setting a tag attribute, e.g.  `tag: "component"`?  Again you don't need that.
3. Why create a closure, but then use `v.state`, you can define normal variables in the closure and reference them [this](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4hcYgAIAYjBgATAEYZqAa0kBeScAA6aSZIBuEGAHdEkgBSGAlJoB8Vw-gzFiAJzj44hWqckA-JJYlgDkoRShACqEGGjqAJ60AK4AhKF2FmjJUFA2FHqsenriUgDCtFgADvQimlZ2Go66+gaSsFJgckoqqlEQOO71AAwA3Hpt7TCd3cpqAMq+pvpaYBhQcDDjrQYTbaVGJv5a1o2Oe5Mh4ReTkhTBYYrJbvShN7d32pL01FAQahZLDBbA5tO8PpJgfgqu5gSIACIwNY5YiWGzbCG3LoKOZ9AYwIZaTbEfo4FKotGglqYj7YnoLJYrSQeZJbcGYrD4WHydwYUxojE0gzsSQAJmGEvR7LarCKOwh91CAAkYLlaJIIkYXG5PPhMDhpXZpfcQrIcb17sBJD4-BY6bjFn59KwbODXWgbrDiMl3PorcYzJI5XKSvQJJIlZUYAAFDAAcxg9WphzMgLOVhuVze8oMJssFWqtQYlsk+pgFlCACFaIpQkGjTnPiECzU0CIS2WKwBBP7UGB1l2SG7ugz5QrbPSc9zkmCWeS0ajJHAMfCKWjyBL3HQgAD0vgNlDBrW3e6j24skZwsYTBTQLu2lBolSq0AJSGyuSom1g1GIEDDeDDIgwwALQAMwABzAWwHAgGWeD4NQcACE+wjMDwbAALpUH88QICgnA4HgWAQMQhDuNAj4+uQPAkMQVRwIgO47skaBVKocaIZUO4kWRFFQAAAqK+DCQAjMMPGkeR0D4CRaD4Pwj7EAkVTcCAcDUBRVSiKwmGsEAA)

I don't know if that helps, but I think if you adapt your source accordingly your bug will probably disappear.

**Comment by EverettMcKay** [comment]
At some point I discovered that if you have an array of mithril components of the same type, Mithril apparently feels free to share state. Assigning a key to these components solves this problem.

Apologies for not closing this earlier.

**Reply by dead-claudia** [reply]
It's not that. It's that Mithril tracks the state identity by position if you don't add keys.

If you do `array.map(thing => m(MyComponent, {thing}))`, the first component is updated with the first `thing`, the second with the second `thing`, and so on. It's simple, but often too simple.

Conversely, if you do `array.map(thing => m(MyComponent, {thing, key: thing.id}))`, it takes each `thing`, moves the previous component with key `thing.id` to the offset of the current `thing`, and updates with that previous component. It's more complicated, but also what you usually need with lists of interactive or animated elements.
